gobblin-cluster/src/main/java/org/apache/gobblin/cluster/GobblinHelixJobLauncher.java

```diff
@@ -462,10 +466,14 @@ public void launchJob(@Nullable JobListener jobListener) throws JobException {
       }
 
       // TODO: Better error handling. The current impl swallows exceptions for jobs that were started by this method call.
-      // One potential way to improve the error handling is to make this error swallowing conifgurable
+      // One potential way to improve the error handling is to make this error swallowing configurable
     } catch (Throwable t) {
       errorInJobLaunching = t;
     } finally {
+      if (isCancelWorkflowOnExitEnabled) {
+        cancelJob(jobListener);
```

Comment:
I'm not getting the idea to put this under the final block and add a flag to determine the logic. Can we consider putting it under the catch block and check the value of isLaunched to determine whether we need to cancel the job or not?

Reply:
My question to you is what reason wouldn't clean up the state on Helix side after the completion of a job? 

As of right now, we are doing all deletion of workflows on startup. When the AM shuts down it doesn't cancel anything on the Helix side (error or not).

Reply:
For example, we always clean up the Gobblin state at the end of the job in the finally block. It's not like there's a point to keeping this helix job around. In the edge case scenario this Helix flow runs, we will cause issues with ZK / Helix

Reply:
Your approach does work though. If we don't want a flag, and GaaS fliptop also wants to cancel the workflow on error (I doubt this is the case), then yes we can do your approach without a flag.

What do you think?

Reply:
If the workflow is complete correctly without exception throw out, doesn't that mean helix already marks the workflow as complete as it's in a clean state? what's the point to call cancel in that case? I believe in ingestion cases, We should never exit with complete, if we don't hit an exception, that means we already mark the workflow as deleted on the helix side and exit gracefully.  Removing work unit states make sense as no matter a workflow is complete or failed or canceled, we should do the housekeeping. 
So I still recommend doing that in the exception block.

Reply:
Addressed in latest iteration to avoid unnecessary delete / cancel calls to Helix as we discussed